change CF-friendly 'recipient_variables' to MG-required 'recipient-variables',  if needed

ericleversen opened the pull request:
I  needed to use recipient-variables for batch sending but mailgun uses the variable name recipient-variables, which is invalid in ColdFusion, so this allows the sendMessage() function to use recipient_variables (with underscore) and then translates it into recipient-variables before sending to Mailgun.

mjclemente commented:
Thanks for the PR! I'm happy to incorporate the functionality that you're looking for, but I'm having a hard time using the `recipient_variables` functionality with the changes you've made.

It could be an issue on my part. Could you provide a code same (maybe a gist), of how this is working for you? Something along the lines of 

```
mailGun.sendMessage(
    from = 'from@from.com',
    to = 'to@example.com',
    subject = 'Recipient Variables Example',
    html = '<h3>Yay variables!</h3><p> %recipient.name%</p>',
    recipient_variables = ??? );
```

Also, do you know if they provide documentation for the `recipient-variables` argument? I've found a few random references to it, but nothing in the primary documentation for sending messages via the API (https://documentation.mailgun.com/en/latest/api-sending.html#sending).

Again, thanks for the PR, and happy to help get this working for you.

ericleversen commented:
Hi, Matthew

That was my first ever PR, so apologies for not providing a code example. Lesson learned.
mailGun.sendMessage(
	from = 'from@from.com’,
	to = 'to@example.com’, 
	subject = 'Recipient Variables Example', html = '<h3>Yay variables!</h3><p> %recipient.name%</p>’,
	recipient_variables=‘{"to@example.com": {“name":”Example Name"}}’)


Here’s MG’s documentation on it:
https://documentation.mailgun.com/en/latest/user_manual.html#batch-sending <https://documentation.mailgun.com/en/latest/user_manual.html#batch-sending>. (Search for 'Recipient Variables’)


Thanks for getting that package put together. It made my life easier.

-Eric

> On Feb 21, 2018, at 7:39 PM, Matthew J. Clemente <notifications@github.com> wrote:
> 
> Thanks for the PR! I'm happy to incorporate the functionality that you're looking for, but I'm having a hard time using the recipient_variables functionality with the changes you've made.
> 
> It could be an issue on my part. Could you provide a code same (maybe a gist), of how this is working for you? Something along the lines of
> 
> mailGun.sendMessage(
>     from = 'from@from.com',
>     to = 'to@example.com',
>     subject = 'Recipient Variables Example',
>     html = '<h3>Yay variables!</h3><p> %recipient.name%</p>',
>     recipient_variables = ??? );
> Also, do you know if they provide documentation for the recipient-variables argument? I've found a few random references to it, but nothing in the primary documentation for sending messages via the API (https://documentation.mailgun.com/en/latest/api-sending.html#sending <https://documentation.mailgun.com/en/latest/api-sending.html#sending>).
> 
> Again, thanks for the PR, and happy to help get this working for you.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/mjclemente/mailgun.cfc/pull/3#issuecomment-367538645>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB2CQz93jwOMnRT5zA24rLV6CUGjb7xWks5tXMVpgaJpZM4SM8sP>.
>

mjclemente commented:
No worries - I remember stressing over my first PR - who am I kidding, I'm still intimidated by them.

Let me take a look at this over the weekend. You're passing in the recipient variables as a string; I think I'd rather it be able to handle either a struct or a valid json string. Might not be possible, with how I have this wrapper set up, but I want to look into it.

I also want to look at how other libraries handle this issue. PHP/Python don't allow hyphens either (I don't think), so I want to look at their approach to recipient variables.

mjclemente commented:
Okay, so I had the chance to take a deeper look at this. A few notes - and I apologize in advance for the length of this:

1. Thank you seeing this issue and taking the time to address it. It's a really helpful feature; I wasn't even aware of it, because it's missing from the [Mailgun docs for sending via the API](https://documentation.mailgun.com/en/latest/api-sending.html#sending). I'm gonna open a PR there to add it.
2. With regard to how to handle it, the other language libraries didn't address the hyphenated `recipient-variables` directly, because their handling of the mail object is different. (Most use a builder object, which is an approach I've taken with other wrappers, and probably the direction that I'll ultimately take for this repo, if I have the time). It's just a lot of rewriting, but the result is nicer.
3. In terms of merging this PR and how I'd like `recipient-variables` handled here, I had a few thoughts:
    1. One is minor - no need for the wrapping `cfscript` tags.
    2. I'd rather see this addressed within `sendMessage()`, instead of having it impact every request, in `apiCall()`.
    3. Ideally, I'd like it to be able to handle structs or JSON strings, but if you didn't include that, I could add that functionality later.

If you want to tackle this, go for it! If you'd rather not, feel free to open an issue on the repo, requesting the functionality, and I'll take a stab at it.

Thanks again!

mjclemente commented:
Hey, just wanted to check to see if you had a chance to look at my response above regarding how we can get this update merged. Thoughts?

ericleversen commented:
I've removed CFscript tags and moved relevant function to sendmessage(). I am not able to do the struct/json request.

I'm new to pull requests...should I create a new one with updated code or add to this one? If latter, how is that done?

mjclemente commented:
Whatever is easier for you if fine with me.

[This article](https://stackoverflow.com/questions/16748115/how-to-modify-github-pull-request) seems to say that if you just push your commits to the branch you made the pull from, the PR will pick them up, so there wouldn't be a need to open a new one.

mjclemente commented:
Any progress on this? (Not trying to be a pain - it's just on a to-do list now, so I'm circling back to it every once in a while.)

Also, I did notice that on the Github page for the PR, it has the (hopefully) helpful tip: 
> Add more commits by pushing to the **recipient-variables-workaround** branch on **ericleversen/mailgun.cfc**.

mjclemente commented:
Haven't heard back from you, so I'm closing this. Please feel free to open a new PR with the fixes you've mentioned. All the best!
